### PR TITLE
Image: add tomcat

### PR DESCRIPTION
**Pull request by eddiezane**
Closes https://github.com/chainguard-dev/image-requests/issues/94

### Pre-review Checklist

- [ ] **IMPORTANT: 'image-request' tag has been applied if this PR is adding any images, including new versions or variants**

#### Quality Requirements for Images PRs
The items in this checklist should all be checked in the PR with exceptions clearly documented.
The general idea is that to the extent possible, the image should be a drop-in replacement to its public counterpart.
- [x] Image is a smaller size than its common public counterpart, or if not, the reason why should be explained
- [ ] Image should be scanned for CVEs and should return 0 or near-0 results. Exceptions should be explained.
- [x] The software in the image should be able to be installed and work as expected
  - [ ] The image should start in a cluster
  - [ ] The application should be accessible to the user/cluster/etc. as expected
- [ ] If there is an open source helm chart provided, the image should be able to be used with the chart. If needed, please add a -compat package to close any gaps with the public helm chart.
- [ ] build for x86_64 and aarch64 if possible, document exceptions

#### For new image PRs only

If you have an apko.yaml file in this PR you need to follow this checklist, otherwise feel free to remove.
- [x] Include tests, sufficient enough that you would trust this image running in production.
- [x] Perform manual tests as needed and document your testing results
  - [ ] For builder images (go, python, etc), build a sample app successfully
  - [x] For services images (rabbit, databases, webservers) test basic functionality, upstream install/getting started, port availability, admin access. Document differences from public image.
  - [ ] Applications intended for use with a specific cloud provider must be tested on that cloud provider.

- [x] The version included is the latest GA version of the software
- [x] The latest tag points to the newest stable version
- [ ] There is a dev tag available that includes a shell and apk tools (by depending on 'wolfi-base')
- [x] The image runs as `nonroot` and GID/UID are set to 65532 or upstream default
  - [ ] Alternatively the username and GID/UID may be a commonly used one from the ecosystem e.g: postgres
- [x] The image contains the ca-certs bundle if needed
- [x] ENTRYPOINT
  - [x] For applications/servers/utilities call main program with no arguments e.g. [redis-server]
  - [ ] For base images leave empty
  - [ ] For dev variants set to entrypoint script that falls back to system
- [x] CMD:
  - [x] For server applications give arguments to start in daemon mode (may be empty)
  - [ ] For utilities/tooling bring up help e.g. `–help`
  - [ ] For base images with a shell, call it e.g. [/bin/sh]

- [x] Add annotations e.g:
```
annotations:
  "org.opencontainers.image.authors": "Chainguard Team https://www.chainguard.dev/"
  "org.opencontainers.image.url": https://edu.chainguard.dev/chainguard/chainguard-images/reference/busybox/ # use the academy site here
  "org.opencontainers.image.source": https://github.com/chainguard-images/images/tree/main/images/bazel # use github here
```
- [ ] Check if environment variables are needed e.g. to set data locations
- [ ] Ensure the image responds to SIGTERM
  - [ ] `docker kill $(docker run -d --rm cgr.dev/chainguard/nginx)`
- [ ] Error logs write to stderr and normal logs to stdout. DO NOT write to file.
#### Documentation Requirements

- [ ] Provide a README file, follow the [README template](readme-template.md) to include variants, tags, and usage examples
- [ ] Document your testing and results in the PR
- [ ] Document any exceptions to the testing and config requirements outlined in this template
- [ ] Document where and how the image deviates from popular alternatives (ports, packages, user, entrypoint, versions, etc). Explain why there's a difference and if the user needs to do anything differently when using the image

### CODE REVIEW

File: images/tomcat/configs/latest.apko.yaml [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+  groups:
+    - groupname: nonroot
+      gid: 65532
+  users:
+    - username: nonroot
+      uid: 65532
+      gid: 65532
+  run-as: nonroot
+
+paths:
```
Comment (eddiezane) [inline images/tomcat/configs/latest.apko.yaml:31]: Should `/usr/share/tomcat` just be owned by the nonroot user?

Have been playing whack-a-mole with these dirs and there could be some that haven't been touched in my testing path.

**Comment by eddiezane** [comment]
```
trivy image tomcat:local-amd64
2023-06-09T09:21:43.145-0600    INFO    Loaded trivy.yaml
2023-06-09T09:21:43.153-0600    INFO    Need to update DB
2023-06-09T09:21:43.153-0600    INFO    DB Repository: ghcr.io/aquasecurity/trivy-db
2023-06-09T09:21:43.153-0600    INFO    Downloading DB...
37.40 MiB / 37.40 MiB [-----------------------------------------------------------------------------------------------------------------------------------] 100.00% 39.74 MiB p/s 1.1s
2023-06-09T09:21:44.809-0600    INFO    Vulnerability scanning is enabled
2023-06-09T09:21:44.809-0600    INFO    Secret scanning is enabled
2023-06-09T09:21:44.809-0600    INFO    If your scanning is slow, please try '--scanners vuln' to disable secret scanning
2023-06-09T09:21:44.809-0600    INFO    Please see also https://aquasecurity.github.io/trivy/v0.42/docs/secret/scanning/#recommendation for faster secret detection
2023-06-09T09:21:44.811-0600    INFO    Detected OS: wolfi
2023-06-09T09:21:44.811-0600    INFO    Detecting Wolfi vulnerabilities...
2023-06-09T09:21:44.811-0600    INFO    Number of language-specific files: 0

tomcat:local-amd64 (wolfi 20230201)

Total: 0 (UNKNOWN: 0, LOW: 0, MEDIUM: 0, HIGH: 0, CRITICAL: 0)
```

```
grype docker:tomcat:local-amd64
 ✔ Vulnerability DB        [updated]
 ✔ Loaded image
 ✔ Parsed image
 ✔ Cataloged packages      [89 packages]
 ✔ Scanning image...       [24 vulnerabilities]
   ├── 0 critical, 7 high, 16 medium, 1 low, 0 negligible
   └── 0 fixed
NAME           INSTALLED  FIXED-IN  TYPE          VULNERABILITY  SEVERITY
tomcat-jdbc    10.1.10              java-archive  CVE-2016-5425  High
tomcat-jdbc    10.1.10              java-archive  CVE-2016-6325  High
tomcat-native  2.0.4                java-archive  CVE-2000-1210  Medium
tomcat-native  2.0.4                java-archive  CVE-2001-0590  Medium
tomcat-native  2.0.4                java-archive  CVE-2002-0493  High
tomcat-native  2.0.4                java-archive  CVE-2005-4838  Medium
tomcat-native  2.0.4                java-archive  CVE-2006-7196  Medium
tomcat-native  2.0.4                java-archive  CVE-2007-1358  Low
tomcat-native  2.0.4                java-archive  CVE-2007-2449  Medium
tomcat-native  2.0.4                java-archive  CVE-2008-0128  Medium
tomcat-native  2.0.4                java-archive  CVE-2009-2696  Medium
tomcat-native  2.0.4                java-archive  CVE-2013-2185  High
tomcat-native  2.0.4                java-archive  CVE-2013-4286  Medium
tomcat-native  2.0.4                java-archive  CVE-2013-4322  Medium
tomcat-native  2.0.4                java-archive  CVE-2013-4444  Medium
tomcat-native  2.0.4                java-archive  CVE-2013-4590  Medium
tomcat-native  2.0.4                java-archive  CVE-2013-6357  Medium
tomcat-native  2.0.4                java-archive  CVE-2014-0075  Medium
tomcat-native  2.0.4                java-archive  CVE-2014-0096  Medium
tomcat-native  2.0.4                java-archive  CVE-2014-0099  Medium
tomcat-native  2.0.4                java-archive  CVE-2014-0119  Medium
tomcat-native  2.0.4                java-archive  CVE-2016-5425  High
tomcat-native  2.0.4                java-archive  CVE-2016-6325  High
tomcat-native  2.0.4                java-archive  CVE-2020-8022  High
```

File: images/tomcat/configs/latest.apko.yaml [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+  - path: /usr/share/tomcat/conf
+    type: directory
+    permissions: 0o755
+    uid: 65532
+    gid: 65532
+  - path: /usr/share/tomcat/logs
```
Comment (ajayk) [inline images/tomcat/configs/latest.apko.yaml:37]: Tomcat writes to logs , can we redirect it to stdout ?

Comment (eddiezane) [reply images/tomcat/configs/latest.apko.yaml:37]: As far as I can tell it looks like the expectation for Kubernetes users is to run a sidecar like fluent.

It does look like this is [possible](https://medium.com/codex/configuring-tomcat-logging-for-a-read-only-file-system-in-kubernetes-965556e55045) though.

I'm waiting to hear back from the customer on what they expect but should this image still switch to logging to stdout?

File: images/tomcat/README.md [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+# tomcat
+| | |
+| - | - |
+| **Status** | stable |
+| **OCI Reference** | `cgr.dev/chainguard/tomcat` |
+
+
+* [View Image in Chainguard Academy](https://edu.chainguard.dev/chainguard/chainguard-images/reference/tomcat/overview/)
+* [View Image Catalog](https://console.enforce.dev/images/catalog) for a full list of available tags.
+*[Contact Chainguard](https://www.chainguard.dev/chainguard-images) for enterprise support, SLAs, and access to older tags.*
```
Comment (tstromberg) [inline images/tomcat/README.md:11]: Should there be a space before the [?

File: images/tomcat/tests/02-run.sh [hunk 4]
```diff
@@ -0,0 +1,23 @@
+#!/usr/bin/env bash
+
+set -o errexit -o nounset -o errtrace -o pipefail -x
+
+if [[ "${IMAGE_NAME}" == "" ]]; then
```
Comment (tstromberg) [inline images/tomcat/tests/02-run.sh:5]: This is more idiomatic:

`if [[ -z "${IMAGE_NAME}" ]]; then` 

(here and elsewhere)

File: images/tomcat/tests/02-run.sh [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+    echo "Must set IMAGE_NAME environment variable. Exiting."
+    exit 1
+fi
+
+TMPDIR="$(mktemp -d)"
+
+function cleanup() {
+    rm -rf "$TMPDIR"
```
Comment (tstromberg) [inline images/tomcat/tests/02-run.sh:13]: Use `"${TMPDIR}"` for consistency.

Comment (tstromberg) [reply images/tomcat/tests/02-run.sh:13]: also, is the -f really necessary?

File: images/tomcat/tests/02-run.sh [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+}
+
+trap cleanup EXIT
+
+curl -sSL -o "${TMPDIR}/sample.war" 'https://tomcat.apache.org/tomcat-10.0-doc/appdev/sample/sample.war'
+
+CONTAINER=$(docker run -v "${TMPDIR}/sample.war:/usr/share/tomcat/webapps/sample.war" -p 8080:8080 -d --rm "${IMAGE_NAME}")
+sleep 3
+curl localhost:8080/sample/ | grep "Hello, World"
```
Comment (tstromberg) [inline images/tomcat/tests/02-run.sh:22]: optional nit: This is racy, so add `--retry 3` for more reliable testing.

**Review by tstromberg — CHANGES_REQUESTED** [review]

**Pull request closed without merging**